pkg/ccl/backupccl/restore_job.go

```diff
@@ -1512,6 +1520,64 @@ func createImportingDescriptors(
 	return dataToPreRestore, dataToRestore, nil
 }
 
+// remapPublicSchemas is used to create a descriptor backed public schema
+// for databases that have virtual public schemas.
+// The rewrite map is updated with the new public schema id.
+func remapPublicSchemas(
+	ctx context.Context,
+	p sql.JobExecContext,
+	mutableDatabases []*dbdesc.Mutable,
+	schemasToWrite *[]*schemadesc.Mutable,
+	writtenSchemas *[]catalog.SchemaDescriptor,
+	details *jobspb.RestoreDetails,
+) error {
+	databaseToPublicSchemaID := make(map[descpb.ID]descpb.ID)
+	// Create descriptor backed Public schema if necessary.
+	for _, db := range mutableDatabases {
+		if db.Schemas == nil || !db.HasPublicSchemaWithDescriptor() {
+			id, err := catalogkv.GenerateUniqueDescID(ctx, p.ExecCfg().DB, p.ExecCfg().Codec)
+			if err != nil {
+				return err
+			}
+
+			db.Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
+				tree.PublicSchema: {
+					ID:      id,
+					Dropped: false,
+				},
+			}
+			// Every database must be initialized with the public schema.
+			// Create the SchemaDescriptor.
+			// In postgres, the user "postgres" is the owner of the public schema in a
+			// newly created db. Postgres and Public have USAGE and CREATE privileges.
+			// In CockroachDB, root is our substitute for the postgres user.
+			publicSchemaPrivileges := descpb.NewBasePrivilegeDescriptor(security.AdminRoleName())
+			// By default, everyone has USAGE and CREATE on the public schema.
+			publicSchemaPrivileges.Grant(security.PublicRoleName(), privilege.List{privilege.CREATE, privilege.USAGE})
+			publicSchemaDesc := schemadesc.NewBuilder(&descpb.SchemaDescriptor{
+				ParentID:   db.GetID(),
+				Name:       tree.PublicSchema,
+				ID:         id,
+				Privileges: publicSchemaPrivileges,
+				Version:    1,
+			}).BuildCreatedMutableSchema()
+
+			*schemasToWrite = append(*schemasToWrite, publicSchemaDesc)
+			*writtenSchemas = append(*writtenSchemas, publicSchemaDesc)
+			databaseToPublicSchemaID[db.GetID()] = id
+		}
+	}
+
+	// Now we need to handle rewriting the table parent schema ids.
+	for id, rw := range details.DescriptorRewrites {
```

Comment:
This does not seem correct. We are iterating over all schema objects that are being restored, and if they are in the database for which we synthesized a public schema descriptor, we are rewriting the parent schema ID to public? Doesn't this put everything in the public schema?

Reply:
Can we be more selective about only setting the `ParentSchemaID` field for types, tables, and sequences, if they were previously mapped to the `PublicSchemaIDForBackup?